### PR TITLE
add a github workflow action to make a release from tags

### PR DESCRIPTION
**Pull request by gdamjan**
make a github (pre)release for every tag that starts with `v*`.

info about the action used:
https://github.com/softprops/action-gh-release

### CODE REVIEW

**Comment by gdamjan** [comment]
main motivation is to have this backported to the systemd-stable repository too, so that we have Github releases on new tags, and people can subscribe to get notified on Github.

For now, I've set the action to make a "prerelease", since the release notes/NEWS currently have been manually edited. But we could also automate that with a step that generates the release notes file, and set it as `body_path` parameter to the action.

Alternatively, `generate_release_notes: true` could (try to) generate the release notes - might be useful for -stable.

**Comment by gdamjan** [comment]
probably should add
```
permissions:
  contents: write
```
depending on the default repo settings

**Comment by bluca** [comment]
So for main, I'd prefer the release to appear (and the notifications to fly off) after the release notes are added, is that possible?
For stable, we could simply add it directory to the stable repository

File: .github/workflows/make_release.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Release
+        uses: softprops/action-gh-release@v1
+        with:
+          prerelease: true
```
Comment (keszybz) [inline .github/workflows/make_release.yml]: Hmm, any chance you could change this to be conditional, and set prerelease if there's `-rc` in the release name? I don't think we want to set the tag to pre-release for final releases. It is fine if we add the notes later, editing an existing tag.

Comment (gdamjan) [reply .github/workflows/make_release.yml]: good call. I'll check it out

Comment (gdamjan) [inline .github/workflows/make_release.yml]: something like this:
```suggestion
          prerelease: ${{ contains(github.ref_name, "-rc") }}
```
should work. "prerelease" will be true if the tag contains "-rc", and false if it doesn't.

Comment (gdamjan) [reply .github/workflows/make_release.yml]: image

**Review by keszybz — COMMENTED** [review]
LGTM.

FWIW, I think if the release appears immediately when the tag is pushed.

**Comment by bluca** [comment]
This looks good for the systemd-table repo branches, and also looks good for here if we plan to do a follow-up and automate pulling NEWS into the release. But if not, for myself I'd prefer to click the button myself, once the release notes are edited, so that the notification goes out once it's all done.

File: .github/workflows/make_release.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Release
+        uses: softprops/action-gh-release@v1
+        with:
+          prerelease: ${{ contains(github.ref_name, '-rc') }}
```
Comment (gdamjan) [inline .github/workflows/make_release.yml:17]: so add this too @bluca ?
```suggestion
          prerelease: ${{ contains(github.ref_name, '-rc') }}
          draft: ${{ github.repository == 'systemd/systemd' }}
```

Comment (bluca) [reply .github/workflows/make_release.yml:17]: yes that would be nice

Comment (gdamjan) [reply .github/workflows/make_release.yml:17]: done

**Comment by evverx** [comment]
> depending on the default repo settings

The actions in the systemd and systemd-stable repositories have read-only access by default so in its current form this action can't upload anything. It isn't clear what exactly was "reviewed" and rubberstamped here.

**Comment by evverx** [comment]
Before I forget, https://github.com/evverx/systemd/actions/runs/4590875978 shows how it fails with read-only tokens. Presumably `contents: write` should fix it but looking at https://github.com/softprops/action-gh-release/pull/234 it seems it was deduced by trial and error.

I created a release at https://github.com/evverx/systemd/releases/tag/v1000 just in case (but I haven't verified what exactly it released). At first glance "🤔  not include valid file" in the logs looks weird:
```
2023-04-02T22:18:56.6812992Z 👩‍🏭 Creating new GitHub release for tag v1000...
2023-04-02T22:18:57.1827177Z 🤔  not include valid file.
2023-04-02T22:18:57.1841114Z 🎉 Release ready at https://github.com/evverx/systemd/releases/tag/v1000
2023-04-02T22:18:57.1978627Z Post job cleanup.
```

**Comment by gdamjan** [comment]
> it seems it was deduced by trial and error.

Github did a lot of changes around the default tokens (and are still doing) - and that's not always best documented. Especially since the defaults for new repos are not the same as the previous defaults.

> At first glance "thinking not include valid file" in the logs looks weird

That's because we don't attach any artifacts to the release - so it's fine.

**Comment by evverx** [comment]
> Github did a lot of changes around the default tokens (and are still doing) - and that's not always best documented.

Agreed and it's pretty sad when it comes to actions releasing stuff consumed by actual people. Anyway the systemd repository has the read-only defaults and this action can't release anything here (or in the systemd-stable repository) until its permissions are specified explicitly (and all the other actions specify them explicitly). Generally in its current form this action isn't consistent with the other actions at all. Also I don't think it should be run on forks.

>  That's because we don't attach any artifacts to the release - so it's fine

I'm not sure why it logs it though. I wonder where tarballs come from? Does the action generate them itself or does it use the GH API and let GitHub generates them?

**Comment by gdamjan** [comment]
> I wonder where tarballs come from? Does the action generate them itself or does it use the GH API and let GitHub generate them?

Github creates those archive tarballs for the Release (not even sure they are stored archives or they generate them on-demand).

**Comment by evverx** [comment]
Got it. I just though that it generated them itself and it was possible to (hypothetically) sign them with Sigstore (or something like that). As far as I can remember the Arch folks were interested in that.

**Comment by evverx** [comment]
For some reason it has just been merged into systemd-stable and if I wasn't clear enough it's just going to fail there with
```
2023-04-02T21:19:30.4378102Z 👩‍🏭 Creating new GitHub release for tag v1000...
2023-04-02T21:19:30.6977516Z ⚠️ GitHub release failed with status: 403
2023-04-02T21:19:30.6977867Z undefined
2023-04-02T21:19:30.6978083Z retrying... (2 retries remaining)
2023-04-02T21:19:30.9448199Z 👩‍🏭 Creating new GitHub release for tag v1000...
2023-04-02T21:19:31.2032500Z ⚠️ GitHub release failed with status: 403
2023-04-02T21:19:31.2032748Z undefined
2023-04-02T21:19:31.2032958Z retrying... (1 retries remaining)
2023-04-02T21:19:31.4839966Z 👩‍🏭 Creating new GitHub release for tag v1000...
2023-04-02T21:19:31.7511042Z ⚠️ GitHub release failed with status: 403
2023-04-02T21:19:31.7511304Z undefined
2023-04-02T21:19:31.7514067Z retrying... (0 retries remaining)
2023-04-02T21:19:31.7514508Z ❌ Too many retries. Aborting...
2023-04-02T21:19:31.7547690Z ##[error]Too many retries.
2023-04-02T21:19:31.7632140Z Post job cleanup.
```

**Comment by keszybz** [comment]
> It isn't clear what exactly was "reviewed" and rubberstamped here.

@evverx The commit was reviewed based on the reading of the patch and tested in gdamjan's repo. No one set up a clone of the systemd organization to test if the permissions are right. This would be very hard to do, and hardly makes sense anyway, since it would be a huge effort. The patch was not "rubberstamped", even if a mistake slipped through. We are not experts in github action logic language so we missed some permission details. And it's hardly the end of the world. IIUC, the patch to fix this should be two lines.

**Comment by evverx** [comment]
> We are not experts in github action logic language so we missed some permission details. And it's hardly the end of the world

@keszybz I didn't say it was the end of the world but given that this is supposed to pull some random code from GitHub and run it with write access to a lot of stuff in the repository: https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#contents I'd expect the review to be a bit more thorough. It's like `pip install` on steroids that can upload systemd releases and push commits.

Having taken a look at the action code it seems it calls stuff like https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#create-a-release and it indeed needs `write: contents` unfortunately. As I said `write: contents` can do much more than that but if it's worth it adding those permissions should make it work.

> can you confirm that's the fix?

I think it should make it work.

**Comment by evverx** [comment]
> The patch was not "rubberstamped", even if a mistake slipped through

It keeps "slipping through" though: https://github.com/systemd/systemd-stable/pull/275, https://github.com/systemd/systemd-stable/pull/276 and https://github.com/systemd/systemd-stable/pull/277. I mean, I'm not surprised that stuff that doesn't work ends up in the systemd-stable repository but it is what it is.

**Pull request merged**